keps/sig-apps/0032-portable-service-definitions.md

```diff
@@ -0,0 +1,157 @@
+
+But, many applications rely on software as a service (SaaS). The reason for this is to push the operational details on to someone else who specializes in that particular service so the application developers and operators can focus on their application and business logic.
+
+The problem is that one cannot deploy the same application in two different environments by two different providers, if the applications leverages services, with the same set of resources. This includes cases where the service being leveraged is common (e.g., MySQL as a Service). This problem limits application portability and sharing (e.g., in open source).
+
+This KEP is looking to solve this problem by providing Kubernetes compatible objects, via CRDs and Secrets, that can be used in many environments by many providers to make working with common services easier. This can be used for services like database (e.g., MySQL, PostgreSQL), DNS, SMTP, and many others.
+
+### Goals
+
+* Provide a common way to request common services (e.g., MySQL)
+* Provide a common means to obtain credentials to use the service
+* Provide a common method to detect which services are available
+* Provide a system that can be implemented for the major public clouds, on-premise clusters, and local cluster (e.g., Docker for Mac)
+
+### Non-Goals
+
+* Provide an out of the box solution for every bespoke service provided by everyone
+* Replace Kubernetes Service Catalog
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Stories
+
+#### Story 1
+
+As a user of Kubernetes, I can query the services I can declaratively request using a Kubernetes native API. For example, using the command `kubectl get crds` where I can see services alongside other resources that can be created.
+
+#### Story 2
+
+As a user of Kubernetes, I can declaratively request an instance of a service using a custom resource. When the service is provided the means to use that service (e.g., credentials in a secret) are provided in a common and consistent manner. The same resource and secret can be used in clusters running in different locations and the way the service is provided may be different.
+
+#### Story 3
+
+As a cluster operator or application operator, I can discover controllers implementing the CRDs and secrets to support the application portability in my cluster.
+
+#### Story 4
+
+As a cluster operator or application operator, I can set default values and provider custom settings for a service. 
+
+### Implementation Details/Notes/Constraints
+
+To solve the two user stories there are two types of Kubernetes resources that can be leveraged.
+
+1. Custom resource definitions (CRDs) can be used to describe a service. The CRDs can be implemented by controllers for different environments and the list of installed CRDs can be queried to see what is supported in a cluster
```

Comment:
Given there could be multiple controllers involved, should there be an annotation or field to use as an identifier (e.g. something akin to the ingress.class annotation)?